### PR TITLE
Jakarta Servlet 4.0

### PR DESCRIPTION
**Pull request by arjantijms**
Signed-off-by: arjantijms <arjan.tijms@gmail.com>

## Specification PR template
When creating a specification project release review, create two PRs with the content divided as follows.

Include the following in PR#1:
- [X] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [X] Specification PDF in the form of wombat_x.y.pdf
- [X] Specification HTML in the form of wombat_x.y.html
- [X] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [x] For a Release Review, a summary that a Compatible Implementation
      is complete, passes the TCK, and that the TCK includes sufficient
      coverage of the specification. The TCK users guide MUST include
      the instructions to run the compatible implementations used to
      validate the release.  Instructions MAY be by reference.
- [x] The URL of the OSSRH staging repository for the api, javadoc:
      https://oss.sonatype.org/content/repositories/staging/jakarta/servlet/jakarta.servlet-api/4.0.3/
 - [x] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
    http://download.eclipse.org/ee4j/jakartaee-tck/jakartaee8-eftl/promoted/eclipse-servlet-tck-4.0.0.zip
- [x] The URL of the compatibility certification request issue:
      https://github.com/eclipse-ee4j/servlet-api/issues/262

### CODE REVIEW

**Comment by dblevins** [comment]
Spec status is Draft, but should be Final or Final Release.

**Comment by starksm64** [comment]
Should the candidate EFTL binary should be linked to somewhere under downloads.eclipse.org/ee4j/servlet?

**Comment by bshannon** [comment]
> Should the candidate EFTL binary should be linked to somewhere under downloads.eclipse.org/ee4j/servlet?

Yes, the link above to oss.sonatype.org is not to the EFTL TCK.

Also, the TCK results link needs to be to a more stable URL, not a Jenkins job.

File: servlet/4.0/_index.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+---
+title: "Jakarta Servlet 4.0"
+date: 2019-07-24T23:21:09+0200
```
Comment (kwsutter) [inline servlet/4.0/_index.md]: It sounds like we can remove the time component from the datestamp...

File: servlet/4.0/_index.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+|Red Hat                |         |         |          |
+|Tomitribe              |         |         |          |
+|EE4J PMC               |         |         |          |
+|Participant Members    |         |         |          |
+|Committer Members      |         |         |          |
+
+
+# Compatible Implementations
+
+* GlassFish 5.1
```
Comment (kwsutter) [inline servlet/4.0/_index.md]: Need a link to the proper CI binaries.

**Comment by dblevins** [comment]
I removed this bullet and added the standard version which is a certification request
```
- [x] The URL of the TCK results summary:
      https://jenkins.eclipse.org/servlet/job/2_servlet-run-tck-against-staged-build/lastSuccessfulBuild/artifact/servlettck/bin/summary.txt/*view*/


```

File: servlet/4.0/_index.md (hunk 2 above)
Comment (edbratt) [reply servlet/4.0/_index.md]: You can use 

[Eclipse GlassFish 5.1.0](https://www.eclipse.org/downloads/download.php?file=/glassfish/glassfish-5.1.0.zip)
--

Comment (edbratt) [reply servlet/4.0/_index.md]: For the Servlet TCK, you can use
[eclipse-servlet-tck-4.0.0.zip](http://download.eclipse.org/ee4j/jakartaee-tck/jakartaee8-eftl/promoted/eclipse-servlet-tck-4.0.0.zip)

File: servlet/4.0/_index.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
+---
+title: "Jakarta Servlet 4.0"
+date: 2019-07-24T23:21:09+0200
+summary: "First release for Jakarta EE 8"
+---
+Jakarta Servlet defines a server-side API for handling HTTP requests and responses.
+
+* [Jakarta Servlet 4.0 Specification Document](./servlet-4.0.pdf) (PDF)
```
Comment (joakime) [inline servlet/4.0/_index.md]: filenames are missing "-spec-" portion should be `servlet-spec-4.0.pdf` and `servlet-spec-4.0.html`

Comment (bshannon) [reply servlet/4.0/_index.md]: I just clarified the checklist to say that "-spec" is preferred but not required.

File: servlet/4.0/_index.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+* [Jakarta Servlet 4.0 Specification Document](./servlet-4.0.html) (HTML)
+* [Jakarta Servlet 4.0 Javadoc](./apidocs)
+* Maven coordinates
+  * [jakarta.servlet:jakarta.servlet-api:jar:4.0.3](https://search.maven.org/artifact/jakarta.servlet/jakarta.servlet-api/4.0.3/jar)
```
Comment (starksm64) [inline servlet/4.0/_index.md]: Need link to TCK

File: servlet/4.0/_index.md (hunk 2 above)
Comment (starksm64) [inline servlet/4.0/_index.md]: Add product link from Ed.

**Comment by starksm64** [comment]
Remove servlet/4.0/images/jakarta_ee_logo_schooner_color_stacked_default.png

**Comment by starksm64** [comment]
Change Version to 4.0 (not 4.0.3) and remove version from spec name.

**Comment by starksm64** [comment]
Copyright appears to be missing from spec files.

**Comment by starksm64** [comment]
Need certification request

**Comment by starksm64** [comment]
For wave 1, this spec should be due by 8/7, so please review this ASAP. Note there is a good checklist for https://github.com/jakartaee/specification-committee/blob/master/spec_review_checklist.md

**Comment by starksm64** [comment]
# Spec Review Checklist

1. Spec PR
  - [x] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [x] Directory of form {spec}/x.y
  - [x] PDF of form {spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [x] HTML of form {spec}-spec-x.y.html ("-spec" preferred but not required)
  - [x] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [x] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://oss.sonatype.org/conte/repositories/staging/jakarta/{spec}/jakarta-{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [ ] Second PR for just apidocs

2. _index.md
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors

File: servlet/4.0/_index.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+title: "Jakarta Servlet 4.0"
+date: 2019-07-24
+summary: "First release for Jakarta EE 8"
+---
+Jakarta Servlet defines a server-side API for handling HTTP requests and responses.
+
+* [Jakarta Servlet 4.0 Specification Document](./servlet-spec-4.0.pdf) (PDF)
+* [Jakarta Servlet 4.0 Specification Document](./servlet-spec-4.0.html) (HTML)
+* [Jakarta Servlet 4.0 Javadoc](./apidocs)
+* [Jakarta Servlet 4.0 TCK](http://download.eclipse.org/ee4j/jakartaee-tck/jakartaee8-eftl/promoted/eclipse-servlet-tck-4.0.0.zip)
```
Comment (bshannon) [inline servlet/4.0/_index.md]: This needs to be the link to the final TCK, not the staged TCK.

**Comment by kwsutter** [comment]
Comments...  (besides the missing checklist items)

- There seems to be some extra files in the doc-files directory of the javadoc.  Maybe leftover from the original Servlet spec?  I don't seem them in use by the current javadoc.

**Comment by starksm64** [comment]
> Comments... (besides the missing checklist items)
> 
> * There seems to be some extra files in the doc-files directory of the javadoc.  Maybe leftover from the original Servlet spec?  I don't seem them in use by the current javadoc.

I created a PR against this PR that fixes the extra files and the javadoc footer issue.

**Comment by starksm64** [comment]
With the latest apidocs merge, this looks good to go.

**Comment by starksm64** [comment]
So 4.0.3 is not exactly the same as 4.0.2, the Java EE 8 release?

**Comment by starksm64** [comment]
That is not correct as the TCKs were supposed to have validated the proposed final api jars that are currently candidates for release in Jakarta EE 8.

**Comment by starksm64** [comment]
4.0.2 was the Java EE 8 release that does not have the updated javadocs, so it really is not valid for the Jakarta EE 8 release.

**Comment by arjantijms** [comment]
To clarify/confirm here as well: 4.0.2 is just what's in the default
GlassFish 5.1.0. It's of course being updated with 4.0.3.

**Comment by arjantijms** [comment]
Btw, the jar that was actually used during the TCK run is archived, and can
be inspected here:

https://jenkins.eclipse.org/servlet/job/2_servlet-run-tck-against-staged-build/lastSuccessfulBuild/artifact/glassfish5/glassfish/modules/jakarta.servlet-api.jar

It has the following pom.properties:

#Generated by Maven
#Tue Aug 13 23:36:40 GMT 2019
version=4.0.3
groupId=jakarta.servlet
artifactId=jakarta.servlet-api

I checked at the bytecode level that this jar, indeed has the @Decprecated
annotation in the ServletRequest class:

@Deprecated
 String getRealPath(String paramString);

**Comment by scottmarlow** [comment]
I don't see "com/sun/ts/tests/signaturetest" in the [tck output](https://jenkins.eclipse.org/servlet/job/2_servlet-run-tck-against-staged-build/lastSuccessfulBuild/artifact/servlet-tck/bin/run.log/*view*/)

Do other spec api tck jobs run the "com/sun/ts/tests/signaturetest" tests?

**Comment by arjantijms** [comment]
> Do other spec api tck jobs run the "com/sun/ts/tests/signaturetest" tests?

Quick check, some do, some don't. It's really a shame that we can't see the job configurations of others jobs and learn from them.

**Comment by arjantijms** [comment]
The JSONB job config was used as an example on the WIKI:

https://wiki.eclipse.org/TCK:ts_jte_pipeline

It looks like the issue is that the servlet.classes are not set here:

https://github.com/eclipse-ee4j/jakartaee-tck/blob/master/docker/servlettck.sh#L61

For JSON-B there are being set:

https://github.com/eclipse-ee4j/jakartaee-tck/blob/master/docker/jsonbtck.sh#L41

**Comment by arjantijms** [comment]
Regarding the javadocs, I noticed it had the date range 2017-2019 in the footer:

servlet-footer

I did a PR to fix this: https://github.com/eclipse-ee4j/servlet-api/pull/265

**Comment by arjantijms** [comment]
PR for updated TCK result: https://github.com/eclipse-ee4j/glassfish/pull/22824

**Comment by arjantijms** [comment]
Updated Javadocs corresponding to newly staged build: https://github.com/jakartaee/specifications/pull/62/commits/fe5d529cd694f9136042e82e85d6e1bee0745c31

**Comment by starksm64** [comment]
I have verified the updated api jar.

**Comment by starksm64** [comment]
* On ballot completion
 - [x] The designated specification committee representative adds this final checklist to the main PR.
 - [ ] The specification committee representative merges the specification PRs.
 - [ ] The designated specification committee representative calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job.
 - [ ] The specification project member who created the staging release promotes the specification api jars to maven central. An example release job script can be found here https://wiki.eclipse.org/MavenReleaseScript.
 - [ ] The EMO updates the specification page with the ballot results.
This list goes on the committed spec index page.
 - [ ] The specification project team should go through the merged spec website page to verify all the links are valid.
 - [ ] The specification project team should approve the compatibility request.
 - [ ] The compatible implementation project/vendor MUST send an email tck@eclipse.org for approval of the compatible implementation for trademark usage.
 - [ ] The specification project team should merge any final release branch as appropriate for the branch management for the project.

**Comment by starksm64** [comment]
@arjantijms you are free to release the api jar to maven central

**Comment by arjantijms** [comment]
> @arjantijms you are free to release the api jar to maven central

@starksm64 I've released the API jar to Maven central: https://repo1.maven.org/maven2/jakarta/servlet/jakarta.servlet-api/4.0.3/

**Comment by bshannon** [comment]
@arjantijms Let us know when the remaining items are complete.

**Comment by arjantijms** [comment]
@bshannon 

Final release branch PR: https://github.com/eclipse-ee4j/servlet-api/pull/268

Can I also sent the TCK email being merely a GF committer, or should the GF project lead do that?

**Comment by bshannon** [comment]
> Can I also sent the TCK email being merely a GF committer, or should the GF project lead do that?

I don't think there are precise rules around that.  As long as you believe you have the support of the GlassFish project lead, I would go ahead and do it.

**Comment by arjantijms** [comment]
Mail sent with @smillidge cc'ed.

**Pull request merged**